### PR TITLE
ido look application module

### PR DESCRIPTION
**Pull request by tomervak**

### CODE REVIEW

File: job-application-tracker-api/src/applications/applications.controller.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+    constructor(private readonly applicationService: ApplicationsService) { }
+
+    @Get()
+    @ApiOkResponse({
+        description: 'list of user applications',
+        type: [ApplicationDto]
+    })
+    async findAllUserApplications(@Request() request: ApiRequest) {
+        if (!request.jwtPayLoad.roles.includes(Role.admin)) throw new BadRequestException('you are not allowed to get this user');
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/applications/applications.controller.ts]: wtf

File: job-application-tracker-api/src/applications/applications.controller.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        return await this.applicationService.findAllUserApplications(request.jwtPayLoad.sub);
+    }
+
+    @Get(':id')
+    @ApiOkResponse({
+        description: 'applicationDto',
+        type: ApplicationDto
+    })
+    async findOne(@Request() request: ApiRequest, @Param('id') id: string) {
+        if (!request.jwtPayLoad.roles.includes(Role.admin) && request.jwtPayLoad.sub !== id) {
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/applications/applications.controller.ts]: wtf

File: job-application-tracker-api/src/applications/applications.controller.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        description: 'applicationDto',
+        type: ApplicationDto
+    })
+    async update(
+        @Request() request: ApiRequest,
+        @Param('id') id: string,
+        @Body(ValidationPipe) updateApplicationDto: UpdateApplicationDto
+    ) {
+        if (request.jwtPayLoad.sub !== updateApplicationDto.userId && !request.jwtPayLoad.roles.includes(Role.admin)) {
+            throw new BadRequestException('id doesnt match');
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/applications/applications.controller.ts]: Not found

File: job-application-tracker-api/src/applications/applications.service.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        catch {
+            throw new NotFoundException('no such user');
+        }
+
+        const newApplication : CreateApplicationDto = {
+            companyName: createApplicationDto.companyName,
+            position: createApplicationDto.position,
+            states: createApplicationDto.states,
+            userId: createApplicationDto.userId
+        };
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/applications/applications.service.ts]: Why would you not send the CreateApplicationDto directly to applicationRepository.create

File: job-application-tracker-api/src/users/users.service.ts [hunk 5]
```diff
@@ -50,24 +49,24 @@ export class UsersService {
...
+        const emailUser = this.userRepository.findByEmail(user.email);
+        if (emailUser) {
+            if ((await emailUser)._id !== applicationId)
                 throw new BadRequestException('this email is being used by a different user');
         }
         if (updateUserDto.password) {
             user.passwordHash = await hash(updateUserDto.password, 10);
         }
-        const updatedUser = await this.userRepository.update(id,user);
+        const updatedUser = await this.userRepository.update(applicationId, user);
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/users/users.service.ts]: Why is this called applicationId

**Review by ido-rosenberger — CHANGES_REQUESTED** [review]

File: job-application-tracker-api/src/applications/dto/create-application.dto.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    @ApiProperty({ example: 'Facebook' })
+    companyName: string;
+
+    @IsString()
+    @IsNotEmpty()
+    @ApiProperty({ example: 'coffe maker' })
+    position: string;
+
+    @ApiProperty()
+    states: any[];
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/applications/dto/create-application.dto.ts]: No need to allow states in CreateApplicationDto (make sure its allways creates as an empty array in the DB)

File: job-application-tracker-api/src/applications/dto/update-application.dto.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    @ApiProperty()
+    companyName: string;
+
+    @IsString()
+    @IsNotEmpty()
+    @ApiProperty()
+    position: string;
+
+    @ApiProperty()
+    states: any[];
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/applications/dto/update-application.dto.ts]: No need to allow modifing the states array in the update request. We will have an API for states

File: job-application-tracker-api/src/applications/dto/update-application.dto.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+    @IsString()
+    @IsNotEmpty()
+    @ApiProperty()
+    userId?: string;
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/applications/dto/update-application.dto.ts]: No need to allow modifing the userId in the update request.

File: job-application-tracker-api/src/applications/applications.service.ts (hunk 4 above)
Comment (tomervak) [reply job-application-tracker-api/src/applications/applications.service.ts]: וואלה לא יודע

File: job-application-tracker-api/src/users/users.service.ts (hunk 5 above)
Comment (tomervak) [reply job-application-tracker-api/src/users/users.service.ts]: באמת שאין לי מושג מה עשיתי פה

Comment (tomervak) [reply job-application-tracker-api/src/users/users.service.ts]: נראלי פשוט התבלבלתי בסרבסים אבל אין לי שמץ

**Comment by tomervak** [comment]
סיימתי

File: job-application-tracker-api/src/applications/applications.controller.ts [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    @Get(':id')
+    @ApiOkResponse({
+        description: 'applicationDto',
+        type: ApplicationDto
+    })
+    async findOne(@Request() request: ApiRequest, @Param('id') id: string) {
+        if (!request.jwtPayLoad.roles.includes(Role.admin)) {
+            throw new BadRequestException('you are not allowed to do this');
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/applications/applications.controller.ts]: Lama

File: job-application-tracker-api/src/applications/applications.controller.ts (hunk 3 above)
Comment (ido-rosenberger) [reply job-application-tracker-api/src/applications/applications.controller.ts]: Are you eeleg?

**Review by ido-rosenberger — CHANGES_REQUESTED** [review]

File: job-application-tracker-api/src/applications/applications.controller.ts [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        type: ApplicationDto
+    })
+    async update(
+        @Request() request: ApiRequest,
+        @Param('id') id: string,
+        @Body(ValidationPipe) updateApplicationDto: UpdateApplicationDto
+    ) {
+        const userId=(await this.applicationService.findOne(updateApplicationDto.id)).userId;
+        if (request.jwtPayLoad.sub!==userId &&!request.jwtPayLoad.roles.includes(Role.admin)) {
+            throw new BadRequestException('Not Found');
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/applications/applications.controller.ts]: Re: Are you eeleg?

Comment (tomervak) [reply job-application-tracker-api/src/applications/applications.controller.ts]: lo hevanti

**Review by ido-rosenberger — APPROVED** [review]

**Pull request merged**